README.md
# NAME

Redis::Jet - Yet another XS implemented Redis Client

# SYNOPSIS

    use Redis::Jet;
    
    my $jet = Redis::Jet->new( server => 'localhost:6379' );
    my $ret = $jet->command(qw/set redis data-server/); # $ret eq 'OK'
    my $value = $jet->command(qw/get redis/); # $value eq 'data-server'
    
    my $ret = $jet->command(qw/set memcached cache-server/);
    my $values = $jet->command(qw/mget redis memcached mysql/);
    # $values eq ['data-server','memcached',undef]
    
    ## error handling
    ($values,$error) = $jet->command(qw/get redis memcached mysql/);
    # $error eq q!ERR wrong number of arguments for 'get' command!

    ## pipeline
    my @values = $jet->pipeline([qw/get redis/],[qw/get memcached/]);
    # \@values = [['data-server'],['cache-server']]

    my @values = $jet->pipeline([qw/get redis/],[qw/get memcached mysql/]);
    # \@values = [['data-server'],[undef,q!ERR wrong...!]]

# DESCRIPTION

This project is still in a very early development stage.
IT IS NOT READY FOR PRODUCTION!

Redis::Jet is yet another XS implemented Redis Client. This module provides
simple interfaces to communicate with Redis server

# METHODS

- `my $obj = Redis::Jet->new(%args)`

    Create a new instance. Can also accept args as hashref.

    - `server => "server:port"`

        server address and port. For using unix\_socket pass socket path to server like \`server => '/path/to/redis.sock'. When server includes \`/\`, Redis::Jet treat server as unix\_socket.

    - connect\_timeout

        Time seconds to wait for establish connection. default: 5

    - io\_timeout

        Time seconds to wait for send request and read response. default: 1

    - utf8

        If enabled, Redis::Jet does encode/decode strings. default: 0 (false)

    - noreply

        IF enabled. The instance does not parse any responses. Every responses to be `"0 but true"`. default: 0 (false)

    - reconnect\_attempts

        If Redis::Jet could not connect to redis server or failed to write requests, Redis::Jet attempts to re-connect. This parameter specify how many times to try reconnect. default: 0 (disable reconnect feature);

    - reconnect\_delay

        Redis::Jet inserts delay before reconnect redis-server (see `reconnect_attempts`). default: 0.1 (100msec)

- `($value,[$error]) = $obj->command($command,$args,$args)`

    send a command and retrieve a value

- `@values = $obj->pipeline([$command,$args,$args],[$command,$args,$args])`

    send several commands and retrieve values. each value has value and error string if error was occurred.

# BENCHMARK

https://gist.github.com/kazeburo/608fc2ba0748e6ab3663

# SEE ALSO

\* [Redis](https://metacpan.org/pod/Redis)

\* [Redis::Fast](https://metacpan.org/pod/Redis%3A%3AFast)

\* [Redis::hiredis](https://metacpan.org/pod/Redis%3A%3Ahiredis)

\* http://redis.io/

# LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

Masahiro Nagano <kazeburo@gmail.com>


Changes
Revision history for Perl extension Redis-Jet

{{$NEXT}}

0.09 2020-03-31T01:11:55Z

   - Supports unix_socket
   - Fix same BIG-KEY segfault bug in pipeline (Thank you itsusony)

0.08 2016-10-05T15:17:30Z

    - Fix a bug with large key and value, improve documents (Thank you itsusony)

0.07 2016-09-30T00:34:04Z

    - Add a positive number validation for correct params (Thank you itsusony)

0.06 2015-11-19T23:57:46Z

    - accept args as hashref in new (thanks adokoy001)

0.05 2015-06-21T02:25:37Z

    - fixed test. changed error message

0.04 2015-06-15T09:01:33Z

    - fixed tests. skip tests if redis-server is not installed

0.03 2015-06-12T16:34:02Z

    - fixed missing context param (Thank you syohex)

0.02 2014-11-07T15:35:05Z

    - refactored
    - bug fix around close socket
    - bug fix around noreply read

0.01 2014-11-04T06:03:50Z

    - original version
      This project is still in a very early development stage.
      IT IS NOT READY FOR PRODUCTION!


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

BEGIN { push @INC, '.' }
use builder::MyBuilder;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl_5',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => '0.4005',
    },

    requires => {
        'perl' => '5.010001',
    },

    recommends => {
    },

    suggests => {
    },

    build_requires => {
    },

    test_requires => {
        'Test::LeakTrace' => '0.15',
        'Test::More' => '0.98',
        'Test::RedisServer' => '0.20',
        'Test::TCP' => '2',
    },

    name            => 'Redis-Jet',
    module_name     => 'Redis::Jet',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = builder::MyBuilder->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();



eg/benchmark.pl
#!/usr/bin/env perl

use strict;
use warnings;
use 5.10.0;
use Benchmark qw/cmpthese/;

use Redis::Fast;
use Redis::Jet;
use Redis;
use Redis::hiredis;

my $redis = Redis->new;
my $hiredis = Redis::hiredis->new(host=>'localhost',port=>'6379');
my $fast = Redis::Fast->new;
my $jet = Redis::Jet->new;
my $jet_noreply = Redis::Jet->new(noreply=>1);

my @val = $jet->command(qw!set foo foovalue!);

print "single get =======\n";

cmpthese(
    -1,
    {
        fast => sub {
            my $val = $fast->get('foo');
        },
        jet => sub {
            my $val = $jet->command(qw/get foo/);
        },
        redis => sub {
            my $data = $redis->get('foo');
        },
        hiredis => sub {
            my $val = $hiredis->command(qw/get foo/);
        },
    }
);

print "single incr =======\n";

cmpthese(
    -1,
    {
        fast => sub {
            my $val = $fast->incr('incrfoo');
        },
        jet => sub {
            my $val = $jet->command(qw/incr incrfoo/);
        },
        jet_noreply => sub {
            $jet_noreply->command(qw/incr incrfoo/);
        },
        redis => sub {
            my $val = $redis->incr('incrfoo');
        },
        hiredis => sub {
            my $val = $hiredis->command(qw/incr incrfoo/);
        },
    }
);

print "pipeline =======\n";

my $cb = sub {};
cmpthese(
    -1,
    {
        fast => sub {
            my @res;
            my $cb = sub { push @res, \@_ };
            $fast->del('user-fail',$cb);
            $fast->del('ip-fail',$cb);
            $fast->lpush('user-log','xxxxxxxxxxx',$cb);
            $fast->lpush('login-log','yyyyyyyyyyy',$cb);
            $fast->wait_all_responses;
        },
        jet => sub {
            my @res = $jet->pipeline(
                [qw/del user-fail/],
                [qw/del ip-fail/],
                [qw/lpush user-log xxxxxxxxxxx/],
                [qw/lpush login-log yyyyyyyyyyy/]
            );
        },
        jet_noreply => sub {
            $jet_noreply->pipeline(
                [qw/del user-fail/],
                [qw/del ip-fail/],
                [qw/lpush user-log xxxxxxxxxxx/],
                [qw/lpush login-log yyyyyyyyyyy/]
            );
        },
        redis => sub {
            my @res;
            my $cb = sub { push @res, \@_ };
            $redis->del('user-fail',$cb);
            $redis->del('ip-fail',$cb);
            $redis->lpush('user-log','xxxxxxxxxxx',$cb);
            $redis->lpush('login-log','yyyyyyyyyyy',$cb);
            $redis->wait_all_responses;
        },
    }
);





lib/Redis/Jet.pm
package Redis::Jet;

use 5.010001;
use strict;
use warnings;
use POSIX qw(EINTR EAGAIN EWOULDBLOCK :sys_wait_h);
use IO::Socket qw(:crlf IPPROTO_TCP TCP_NODELAY);
use IO::Socket::INET;
use IO::Socket::UNIX;
use IO::Select;
use Time::HiRes qw/time/;
use base qw/Exporter/;

our @EXPORT_OK = qw/
                    parse_message parse_message_utf8
                /;

our $VERSION = "0.09";

use XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

sub new {
    my $class = shift;
    my %args = ref $_[0] ? %{$_[0]} : @_;
    %args = (
        server => '127.0.0.1:6379',
        connect_timeout => 5,
        io_timeout => 1,
        utf8 => 0,
        noreply => 0,
        reconnect_attempts => 0,
        reconnect_delay => 0.1,
        %args,
    );
    $class->_new(\%args);
}

sub set_socket {
    my ($self,$socket) = @_;
    $self->get_bucket->{socket} = $socket;
    $self->set_fileno(fileno($socket));
}

sub connect {
    my $self = shift;
    my $socket;
    if ($self->get_server =~ m!/!) {
        $socket = IO::Socket::UNIX->new(
            Peer => $self->get_server,
            Timeout => $self->get_connect_timeout,
        ) or return;
    } else {
        $socket = IO::Socket::INET->new(
            PeerAddr => $self->get_server,
            Timeout => $self->get_connect_timeout,
        ) or return;
        $socket->setsockopt(IPPROTO_TCP, TCP_NODELAY, 1)
            or die "setsockopt(TCP_NODELAY) failed:$!";
    }
    $socket->blocking(0) or die $!;
    $self->set_socket($socket);
    $socket;
}

sub DESTROY {
    $_[0]->_destroy();
}


1;
__END__

=encoding utf-8

=head1 NAME

Redis::Jet - Yet another XS implemented Redis Client

=head1 SYNOPSIS

    use Redis::Jet;
    
    my $jet = Redis::Jet->new( server => 'localhost:6379' );
    my $ret = $jet->command(qw/set redis data-server/); # $ret eq 'OK'
    my $value = $jet->command(qw/get redis/); # $value eq 'data-server'
    
    my $ret = $jet->command(qw/set memcached cache-server/);
    my $values = $jet->command(qw/mget redis memcached mysql/);
    # $values eq ['data-server','memcached',undef]
    
    ## error handling
    ($values,$error) = $jet->command(qw/get redis memcached mysql/);
    # $error eq q!ERR wrong number of arguments for 'get' command!

    ## pipeline
    my @values = $jet->pipeline([qw/get redis/],[qw/get memcached/]);
    # \@values = [['data-server'],['cache-server']]

    my @values = $jet->pipeline([qw/get redis/],[qw/get memcached mysql/]);
    # \@values = [['data-server'],[undef,q!ERR wrong...!]]

=head1 DESCRIPTION

This project is still in a very early development stage.
IT IS NOT READY FOR PRODUCTION!

Redis::Jet is yet another XS implemented Redis Client. This module provides
simple interfaces to communicate with Redis server

=head1 METHODS

=over 4

=item C<< my $obj = Redis::Jet->new(%args) >>

Create a new instance. Can also accept args as hashref.

=over 4

=item C<< server => "server:port" >>

server address and port. For using unix_socket pass socket path to server like `server => '/path/to/redis.sock'. When server includes `/`, Redis::Jet treat server as unix_socket.

=item connect_timeout

Time seconds to wait for establish connection. default: 5

=item io_timeout

Time seconds to wait for send request and read response. default: 1

=item utf8

If enabled, Redis::Jet does encode/decode strings. default: 0 (false)

=item noreply

IF enabled. The instance does not parse any responses. Every responses to be C<"0 but true">. default: 0 (false)

=item reconnect_attempts

If Redis::Jet could not connect to redis server or failed to write requests, Redis::Jet attempts to re-connect. This parameter specify how many times to try reconnect. default: 0 (disable reconnect feature);

=item reconnect_delay

Redis::Jet inserts delay before reconnect redis-server (see C<reconnect_attempts>). default: 0.1 (100msec)

=back

=item C<< ($value,[$error]) = $obj->command($command,$args,$args) >>

send a command and retrieve a value

=item C<< @values = $obj->pipeline([$command,$args,$args],[$command,$args,$args]) >>

send several commands and retrieve values. each value has value and error string if error was occurred.

=back

=head1 BENCHMARK

https://gist.github.com/kazeburo/608fc2ba0748e6ab3663

=head1 SEE ALSO

* L<Redis>

* L<Redis::Fast>

* L<Redis::hiredis>

* http://redis.io/

=head1 LICENSE

Copyright (C) Masahiro Nagano.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=cut

